dowhy/causal_refuters/dummy_outcome_refuter.py

```diff
@@ -98,48 +98,47 @@ def refute_estimate(self):
+        saved_estimator_dict = {}
         
-        # self._scale *= np.std(new_outcome)
-
-        # if len(new_outcome.shape) == 2 and \
-        #     ( new_outcome.shape[0] ==1 or new_outcome.shape[1] ):
-        #     self.logger.warning("Converting the row or column vector to 1D array")
-        #     new_outcome = new_outcome.ravel()
-        #     assert len(new_outcome) == num_rows, ("The number of outputs do not match that of the number of outcomes")
-        # elif len(new_outcome.shape) == 1:
-        #     assert len(new_outcome) == num_rows, ("The number of outputs do not match that of the number of outcomes")
-        # else:
-        #     raise Exception("Type Mismatch: The outcome is one dimensional, but the output has the shape:{}".format(new_outcome.shape))
-
+        X = self._data[self._chosen_variables]
+        new_outcome = self._data['y']
         
-        pdb.set_trace()
         for index in range(self._num_simulations):
-            for action, props in self._pipeline:
-                if props['input'] == ['X']:
-                    new_outcome = action( self._data[self._chosen_variables] )
-                elif props['input'] == ['y']:
-                    new_outcome = action( new_outcome )
-                elif props['input'] == ['X','y']:
-                    temp_estimator = action( new_outcome )
-                    new_outcome = temp_estimator( self._data[self._chosen_variables] )   
-
+            for action, func_args in self._transformations:
+                transform_num = 0
+
+                if callable(action):
+                    new_outcome = action(X, **func_args)
+                elif action in DummyOutcomeRefuter.SUPPORTED_ESTIMATORS:
+                    if save_estimators:
+                        estimator = self._estimate_dummy_outcome(func_args, action, new_outcome)
+                        saved_estimator_dict[action + str(transform_num)] = estimator
+                    
+                    if action + str(transform_num) in saved_estimator_dict:
+                        estimator = saved_estimator_dict[action + str(transform_num)]
+                        new_outcome = estimator(X)
+                    else:
+                        estimator = self._estimate_dummy_outcome(func_args, action, new_outcome)
+                        new_outcome = estimator(X)
+
+                elif action == 'noise':
+                    save_estimators = False
+                    new_outcome = self._noise(new_outcome, func_args)
+                elif action == 'permute':
+                    save_estimators = False
+                    new_outcome = self._permute(new_outcome, func_args)
+                elif action =='zero':
+                    save_estimators = False
+                    new_outcome = np.zeros(new_outcome.shape)
+            
+            transform_num += 1
+            save_estimators = False       

```

Comment:
save_estimators should be made True here, not False, so that the next simulation can begin.

Reply:
Once, we have saved the estimates after the first iteration, we need not run the estimation functions again, as they have been saved in the ```saved_estimator_dict``` . Thus, after the first iteration the value of ```save_estimators``` is made ```False```.

Reply:
got it